### PR TITLE
Added Image capabilities

### PR DESCRIPTION
**Pull request by kwettermann**
Added new procedures for rotating/flipping and clearing an image.

### CODE REVIEW

File: Modules/System Tests/Image/src/ImageTests.Codeunit.al [hunk 1]
```diff
@@ -159,6 +181,32 @@ codeunit 135135 "Image Tests"
...
+        RotateFlipType: Enum "Rotate Flip Type";
+        OutStream: OutStream;
+        ImageAsBase64HorizontalTxt: Label 'iVBORw0KGgoAAAANSUhEUgAAAAoAAAAFCAIAAADzBuo/AAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAAEnQAABJ0Ad5mH3gAAAAUSURBVBhXY6AS+P//PxqDQsDAAADP2QX7LebCcQAAAABJRU5ErkJggg==', Locked = true;
+        ImageAsBase64VerticalTxt: Label 'iVBORw0KGgoAAAANSUhEUgAAAAUAAAAKCAIAAADzWwNnAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAAEnQAABJ0Ad5mH3gAAAAXSURBVBhXYyAP/P//H8oiDpCqHgoYGADezwX75D+gjQAAAABJRU5ErkJggg==', Locked = true;
+    begin
+        // [Given] base64 encoded data, create image
+        Image.FromBase64(ImageAsBase64HorizontalTxt);
+
+        // [When] rotate image
+        Image.RotateFlip(RotateFlipType::Rotate90FlipNone);
```
Comment (mazhelez) [inline Modules/System Tests/Image/src/ImageTests.Codeunit.al:197]: Add a test that rotates an image 4 times and verifies the new image is the same ;)

**Review by mazhelez — CHANGES_REQUESTED** [review]
Great addition!

A few more tests and you're there!

File: Modules/System Tests/Image/src/ImageTests.Codeunit.al [hunk 2]
```diff
@@ -159,6 +181,32 @@ codeunit 135135 "Image Tests"
...
+
+        // [Then] verify image
+        Assert.AreEqual(5, Image.GetWidth(), 'Incorrect width');
+        Assert.AreEqual(10, Image.GetHeight(), 'Incorrect height');
+        Assert.AreEqual(Image.ToBase64(), ImageAsBase64VerticalTxt, 'RotateFlip failed');
```
[the remark this answers is not in the document]
Comment (kwettermann) [reply Modules/System Tests/Image/src/ImageTests.Codeunit.al]: Fixed.

File: Modules/System/Image/src/ImageImpl.Codeunit.al [hunk 3]
```diff
@@ -25,6 +25,37 @@ codeunit 3970 "Image Impl."
...
+
+        TempBlob.CreateInStream(InStream);
+        Image := Image.FromStream(InStream);
+
+        Rectangle := Rectangle.Rectangle(0, 0, CurrentWidth, CurrentHeight);
+
+        BitmapDst := BitmapDst.Bitmap(CurrentWidth, CurrentHeight);
+        Graphics := Graphics.FromImage(BitmapDst);
+
+        Color.FromArgb(255, Red, Green, Blue);
```
[the remark this answers is not in the document]
Comment (kwettermann) [reply Modules/System/Image/src/ImageImpl.Codeunit.al]: Fixed.

File: Modules/System Tests/Image/src/ImageTests.Codeunit.al [hunk 4]
```diff
@@ -124,6 +124,28 @@ codeunit 135135 "Image Tests"
...
+        // [When] Clearing image
+        Image.Clear(255, 0, 0);
+
+        // [Then] verify image
+        Assert.AreEqual(Image.ToBase64(), ImageAsBase64ClearTxt, 'Clear failed');
+
+        // [Then] save to stream
+        TempBlob.CreateOutStream(OutStream);
+        Image.Save(OutStream);
+        Assert.AreNotEqual(0, TempBlob.Length(), 'Image was not saved');
```
[the remark this answers is not in the document]
Comment (kwettermann) [reply Modules/System Tests/Image/src/ImageTests.Codeunit.al:151]: Fixed.

File: Modules/System/Image/src/ImageImpl.Codeunit.al [hunk 5]
```diff
@@ -197,6 +228,17 @@ codeunit 3970 "Image Impl."
...
 
+    procedure RotateFlip(RotateFlipType: Enum "Rotate Flip Type")
+    var
+        Image: DotNet Image;
+        InStream: InStream;
+    begin
+        TempBlob.CreateInStream(InStream);
+        Image := Image.FromStream(InStream);
+        Image.RotateFlip(RotateFlipType.AsInteger());
+        EncodeToImage(Image);
```
[the remark this answers is not in the document]
Comment (kwettermann) [reply Modules/System/Image/src/ImageImpl.Codeunit.al:252]: Yes, this works.

File: Modules/System Tests/Image/src/ImageTests.Codeunit.al [hunk 6]
```diff
@@ -124,6 +124,28 @@ codeunit 135135 "Image Tests"
...
+    var
+        TempBlob: Codeunit "Temp Blob";
+        OutStream: OutStream;
+        ImageAsBase64ClearTxt: Label 'iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAYAAACNMs+9AAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAADsIAAA7CARUoSoAAAAAcSURBVChTY/zPwABEhAETlCYIRhXiBUQqZGAAAD2cAhKD+AbjAAAAAElFTkSuQmCC', Locked = true;
+    begin
+        // [Given] base64 encoded data, create image
+        Image.FromBase64(ImageAsBase64Txt);
+
+        // [When] Clearing image
+        Image.Clear(255, 0, 0);
```
[the remark this answers is not in the document]
Comment (kwettermann) [reply Modules/System Tests/Image/src/ImageTests.Codeunit.al:141]: InvalidClearTest() added.

File: Modules/System/Image/src/Image.Codeunit.al [hunk 7]
```diff
@@ -13,6 +13,17 @@ codeunit 3971 Image
     var
         ImageImpl: Codeunit "Image Impl.";
 
+    /// <summary>
+    /// Clears the entire drawing surface and fills it with the specified background color.
+    /// </summary>
+    /// <param name="Red">Red value.</param>
```
[the remark this answers is not in the document]
Comment (kwettermann) [reply Modules/System/Image/src/Image.Codeunit.al]: Fixed.

File: Modules/System Tests/Image/src/ImageTests.Codeunit.al (hunk 1 above)
Comment (kwettermann) [reply Modules/System Tests/Image/src/ImageTests.Codeunit.al:216]: Added RotateFlip360Test().

**Comment by mazhelez** [comment]
> Hi @mazhelez!
> 
> Thank you for your code review.
> 
> I have updated the code as suggested. Let me know if there are more areas that need an adjustment.

Awesome! I'll take a look shortly 👍

**Comment by mazhelez** [comment]
Hey @kwettermann , there are the following failures:

```
Failure - Codeunit Image Tests -  Duration 0.56 seconds
   Success - Test method: CreateImageFromStreamTest - Duration 0.313 seconds)
   Success - Test method: CreateImageFromBase64Test - Duration 0.01 seconds)
   Success - Test method: FailToCreateImageTest - Duration 0.033 seconds)
   Success - Test method: FailToCreateBase64ImageTest - Duration 0.007 seconds)
   Success - Test method: ConvertToBase64Test - Duration 0.01 seconds)
   Success - Test method: GetFormatTest - Duration 0.014 seconds)
   Success - Test method: GetFormatAsTextTest - Duration 0.003 seconds)
   Success - Test method: GetDimensionsTest - Duration 0.006 seconds)
   Failure - Test method: ClearTest - Duration 0.033 seconds
      Error:
         Assert.AreEqual failed. Expected:<iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAYAAACNMs+9AAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAADsIAAA7CARUoSoAAAAAcSURBVChTY/zPwABEhAETlCYIRhXiBUQqZGAAAD2cAhKD+AbjAAAAAElFTkSuQmCC> (Text). Actual:<iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAYAAACNMs+9AAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAADsMAAA7DAcdvqGQAAAAqSURBVChTjcixDQAACMOw/v90YUYeiOQl2frEKZzCKZzCKZzCKZzCeaQDNL1jnUcuy0wAAAAASUVORK5CYII=> (Text). Clear failed.
      Call Stack:
         Library Assert(CodeUnit 130002).AreEqual line 3 - Library Assert by Microsoft
         Image Tests(CodeUnit 135135).ClearTest line 18 - Image Tests by Microsoft
         Test Runner - Mgt(CodeUnit 130454).RunTests line 18 - Test Runner by Microsoft
         Test Runner - Isol. Codeunit(CodeUnit 130450).OnRun(Trigger) line 4 - Test Runner by Microsoft
         Test Suite Mgt.(CodeUnit 130456).RunTests line 2 - Test Runner by Microsoft
         Test Suite Mgt.(CodeUnit 130456).RunSelectedTests line 34 - Test Runner by Microsoft
         Test Suite Mgt.(CodeUnit 130456).RunNextTest line 16 - Test Runner by Microsoft
         Command Line Test Tool(Page 130455).RunNextTest - OnAction(Trigger) line 7 - Test Runner by Microsoft
         
   Success - Test method: InvalidClearTest - Duration 0.01 seconds)
   Success - Test method: ResizeTest - Duration 0.016 seconds)
   Success - Test method: InvalidResizeTest - Duration 0.003 seconds)
   Failure - Test method: RotateFlipTest - Duration 0.01 seconds
      Error:
         Assert.AreEqual failed. Expected:<iVBORw0KGgoAAAANSUhEUgAAAAUAAAAKCAIAAADzWwNnAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAAEnQAABJ0Ad5mH3gAAAAXSURBVBhXYyAP/P//H8oiDpCqHgoYGADezwX75D+gjQAAAABJRU5ErkJggg==> (Text). Actual:<iVBORw0KGgoAAAANSUhEUgAAAAUAAAAKCAIAAADzWwNnAAAABGdBTUEAALGPC/xhBQAAAAlwSFlzAAAScwAAEnMBjCK5BwAAABdJREFUGFdjIA/8//8fyiIOkKoeChgYAN7PBfvkP6CNAAAAAElFTkSuQmCCAABJRU5ErkJggpjhHgC1mxenYUkDrgAAAABJRU5ErkJggg==> (Text). RotateFlip failed.
      Call Stack:
         Library Assert(CodeUnit 130002).AreEqual line 3 - Library Assert by Microsoft
         Image Tests(CodeUnit 135135).RotateFlipTest line 20 - Image Tests by Microsoft
         Test Runner - Mgt(CodeUnit 130454).RunTests line 18 - Test Runner by Microsoft
         Test Runner - Isol. Codeunit(CodeUnit 130450).OnRun(Trigger) line 4 - Test Runner by Microsoft
         Test Suite Mgt.(CodeUnit 130456).RunTests line 2 - Test Runner by Microsoft
         Test Suite Mgt.(CodeUnit 130456).RunSelectedTests line 34 - Test Runner by Microsoft
         Test Suite Mgt.(CodeUnit 130456).RunNextTest line 16 - Test Runner by Microsoft
         Command Line Test Tool(Page 130455).RunNextTest - OnAction(Trigger) line 7 - Test Runner by Microsoft
         
   Failure - Test method: RotateFlip360Test - Duration 0.007 seconds
      Error:
         Assert.AreEqual failed. Expected:<iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAIAAAACUFjqAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAAJcEhZcwAADsMAAA7DAcdvqGQAAAAOSURBVChTYxgFJAMGBgABNgABY8OiGAAAAABJRU5ErkJggg==> (Text). Actual:<iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAIAAAACUFjqAAAABGdBTUEAALGPC/xhBQAAAAlwSFlzAAAOwQAADsEBuJFr7QAAAA5JREFUKFNjGAUkAwYGAAE2AAFjw6IYAAAAAElFTkSuQmCCGAAAAABJRU5ErkJggk5ErkJggpjhHgC1mxenYUkDrgAAAABJRU5ErkJggg==> (Text). RotateFlip failed.
      Call Stack:
         Library Assert(CodeUnit 130002).AreEqual line 3 - Library Assert by Microsoft
         Image Tests(CodeUnit 135135).RotateFlip360Test line 14 - Image Tests by Microsoft
         Test Runner - Mgt(CodeUnit 130454).RunTests line 18 - Test Runner by Microsoft
         Test Runner - Isol. Codeunit(CodeUnit 130450).OnRun(Trigger) line 4 - Test Runner by Microsoft
         Test Suite Mgt.(CodeUnit 130456).RunTests line 2 - Test Runner by Microsoft
         Test Suite Mgt.(CodeUnit 130456).RunSelectedTests line 34 - Test Runner by Microsoft
         Test Suite Mgt.(CodeUnit 130456).RunNextTest line 16 - Test Runner by Microsoft
         Command Line Test Tool(Page 130455).RunNextTest - OnAction(Trigger) line 7 - Test Runner by Microsoft
```

Can you take a look?

**Comment by kwettermann** [comment]
I have fixed the ClearTest() and RotateFlipTest(). Unfortunately the RotateFlip360Test() still fails. The 360 degree rotation is correct, the images are identical, but the Base64 string differs.

**Comment by kwettermann** [comment]
> > I have fixed the ClearTest() and RotateFlipTest(). Unfortunately the RotateFlip360Test() still fails. The 360 degree rotation is correct, the images are identical, but the Base64 string differs.
> 
> Interesting! Do you think you can figure it out yourself? I can try and help but probably next week.

I think so. I am currently checking the image properties in reference to the EncodeToImage() procedure.

File: Modules/System/Image/src/ImageImpl.Codeunit.al [hunk 8]
```diff
@@ -14,17 +14,53 @@ codeunit 3970 "Image Impl."
...
 
+    procedure Clear(Alpha: Integer; Red: Integer; Green: Integer; Blue: Integer)
+    var
+        Image: DotNet Image;
+        Graphics: DotNet Graphics;
+        Color: DotNet Color;
+        InStream: InStream;
+        CurrentWidth, CurrentHeight : Integer;
+    begin
+        CurrentWidth := GetWidth();
```
[the remark this answers is not in the document]
Comment (mazhelez) [reply Modules/System/Image/src/ImageImpl.Codeunit.al]: If these line stay, make sure to place them after the input validation.

File: Modules/System/Image/src/ImageImpl.Codeunit.al [hunk 9]
```diff
@@ -14,17 +14,53 @@ codeunit 3970 "Image Impl."
...
+        CurrentHeight := GetHeight();
+
+        if (Alpha < 0) or (Alpha > 255) then
+            Error(AlphaErr);
+        if (Red < 0) or (Red > 255) then
+            Error(RedErr);
+        if (Green < 0) or (Green > 255) then
+            Error(GreenErr);
+        if (Blue < 0) or (Blue > 255) then
```
Comment (mazhelez) [inline Modules/System/Image/src/ImageImpl.Codeunit.al:45]: Might look better if you extract this in a separate function.

File: Modules/System/Image/src/ImageImpl.Codeunit.al (hunk 8 above)
[the remark this answers is not in the document]
Comment (kwettermann) [reply Modules/System/Image/src/ImageImpl.Codeunit.al]: No, I have already removed CurrentWidth and CurrentHeight.

**Comment by mazhelez** [comment]
@kwettermann I think I found the issue.

It's an encoding issue, I believe.
I modified the `RotateFlip` to look like
```
    procedure RotateFlip(RotateFlipType: Enum "Rotate Flip Type")
    var
        Image: DotNet Image;
        BitmapDst: DotNet Bitmap;
        InStream: InStream;
    begin
        TempBlob.CreateInStream(InStream);
        Image := Image.FromStream(InStream);

        BitmapDst := BitmapDst.Bitmap(Image);
        BitmapDst.RotateFlip(RotateFlipType.AsInteger());

        EncodeToImage(BitmapDst);
    end;
```

The `RotateFlip360Test` was still failing.., but if we first rotate the image, take the Base64 and then rotate it 4 more times, the tests passes.
```
    [Test]
    procedure RotateFlip360Test()
    var
        Image: Codeunit Image;
        TempBlob: Codeunit "Temp Blob";
        RotateFlipType: Enum "Rotate Flip Type";
        OutStream: OutStream;
        CurrentWidth, CurrentHeight : Integer;
        Original, Rotated : Text;
    begin
        // [Given] base64 encoded data, create image
        Image.FromBase64(ImageAsBase64Txt);
        CurrentWidth := Image.GetWidth();
        CurrentHeight := Image.GetHeight();

        // [When] rotate image 360 degrees
        Image.RotateFlip(RotateFlipType::Rotate90FlipNone);
        Original := Image.ToBase64();

        Image.RotateFlip(RotateFlipType::Rotate90FlipNone);
        Image.RotateFlip(RotateFlipType::Rotate90FlipNone);
        Image.RotateFlip(RotateFlipType::Rotate90FlipNone);
        Image.RotateFlip(RotateFlipType::Rotate90FlipNone);

        Rotated := Image.ToBase64();

        // [Then] verify image
        Assert.AreEqual(CurrentHeight, Image.GetWidth(), 'Incorrect width');
        Assert.AreEqual(CurrentWidth, Image.GetHeight(), 'Incorrect height');
        Assert.AreEqual(Original, Rotated, 'RotateFlip failed');

        // [Then] save to stream
        TempBlob.CreateOutStream(OutStream);
        Image.Save(OutStream);
        Assert.AreNotEqual(0, TempBlob.Length(), 'Image was not saved');
    end;
```

Can you take over from here?

**Comment by kwettermann** [comment]
That is great. Thank you!

I have implemented the changes as you suggested.

**Comment by mazhelez** [comment]
> That is great. Thank you!
> 
> I have implemented the changes as you suggested.

Hey, test were still failing and with [my latest commit](https://github.com/microsoft/ALAppExtensions/pull/14672/commits/ebd74b0ace7421df9453694008187aab0a1ebd8d), they are fixed.

I ended up using a separate Image variable and reused the same image for the tests.
Check it out and let me know what you think

**Comment by mazhelez** [comment]
Hey @kwettermann 

Have a final look, please 😄 
Today I engaged one of our writers to verify the newly added strings. Overall, the PR is ready to be merged.

**Comment by kwettermann** [comment]
I will take a look at it today. :-)

**Comment by kwettermann** [comment]
Hey @mazhelez

The code looks good and the tests passed on my end too.
You can go ahead. 😊 

Thank you so much for your assistance!

**Review by mazhelez — APPROVED** [review]
All good!

I'll soon merge the PR. Changes will be available in version 19.2

**Pull request merged**